warehouse/query-core/src/main/java/datawave/query/common/grouping/Groups.java

```diff
@@ -0,0 +1,95 @@
+package datawave.query.common.grouping;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Represents a set of groups found during a #GROUP_BY operation.
+ */
+public class Groups {
+    
+    /**
+     * A map of distinct grouping values to their groups.
+     */
+    private final Map<Set<GroupingAttribute<?>>,Group> groups = new HashMap<>();
```

Comment:
This is something I've also wanted to improve upon, it's how the original code handled finding groupings. I'll work on coming up with a more efficient key object to use instead of a set.

Reply:
I approved, but I just want to double check that we are going to proceed with a set as the key for this map.  Just wondering where your thought process lead you with this one.

Reply:
So I found myself going back and forth on this, trying to figure out a workable solution that wasn't effectively just a wrapper around a Set. One approach we could try is an immutable HashSet implementation that computes the hashcode once at instantiation and using that as the map key. Thoughts?

Reply:
@jwomeara I pushed a commit that optimizes the hashcode() usage, please take a look if you get a chance and let me know what you think.